### ISSUE TITLE
Function.getValueOpt is Not Fully Implemented

### ISSUE THREAD

**Issue by giovaniceotto** [giovaniceotto/RocketPy#12]
The Function.getValueOpt method is not implemented for the following interpolation methods:

- polynomial
- shepard

Furthermore, it can be made faster by making used of Numba's just in time compiling functions.

**Comment by giovaniceotto** [comment]
Working on this in the [function_arithmetic_operations](https://github.com/giovaniceotto/RocketPy/tree/functions_arithmetic_operations) branch.

**Comment by giovaniceotto** [comment]
The new implementation will include the following features:

- Support for callable, spline, akima, linear, polynomial and shepard.
- Separate method for spline, akima, linear, polynomial and shepard.
- Numba JIT for spline, akima, linear and polynomial.

The use of Numba's JIT make's it much harder to implement the current "remember the last interpolation interval" functionality. However, even without this feature, it can be considerably faster, as can be seen in the following test, where .getValueOpt is the new implementation and .getValueOpt2 is the old one:

```
>> xData = np.linspace(0, 1, 100)
>> yData = xData - xData**2 + xData**3
>> G = Function(np.array([xData, yData]).T, interpolation='spline')
```

Comparison evaluating at a single point:

```
>> %timeit G(0.501)
19.6 µs ± 928 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)

>> %timeit G.getValue(0.501)
19.1 µs ± 1.15 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)

>> %timeit G.getValueOpt(0.501)
The slowest run took 15.00 times longer than the fastest. This could mean that an intermediate result is being cached.
1.12 µs ± 1.39 µs per loop (mean ± std. dev. of 7 runs, 1 loop each)

>> %timeit G.getValueOpt2(0.501)
7.38 µs ± 87.5 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
```

Comparison taking advantage of "remember the last interpolation interval" functionality:

```
>> %timeit [G(i) for i in np.linspace(-1, 2, 1000)]
15.1 ms ± 67.2 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

>> %timeit [G.getValue(i) for i in np.linspace(-1, 2, 1000)]
14.7 ms ± 438 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

>> %timeit [G.getValueOpt(i) for i in np.linspace(-1, 2, 1000)]
437 µs ± 4.45 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)

>> %timeit [G.getValueOpt2(i) for i in np.linspace(-1, 2, 1000)]
8.82 ms ± 46.2 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```

In summary:

- new Function.getValueOpt is 20x faster than the old implementation when successive points are evaluated.
- new Function.getValueOpt is 7x faster than the old implementation when a single point is evaluated many times.

In practice, the new implementation resulted in a 25% speed up in the Flight simulation.
